aptos-move/framework/aptos-framework/sources/Voting.move

```diff
@@ -0,0 +1,286 @@
+        let voting_forum = VotingForum<ProposalType> {
+            next_proposal_id: 0,
+            proposals: Table::new<u64, Proposal<ProposalType>>(),
+            events: VotingEvents {
+                create_proposal_events: Event::new_event_handle<CreateProposalEvent>(account),
+                register_forum_events: Event::new_event_handle<RegisterForumEvent>(account),
+                vote_events: Event::new_event_handle<VoteEvent>(account),
+            }
+        };
+
+        Event::emit_event<RegisterForumEvent>(
+            &mut voting_forum.events.register_forum_events,
+            RegisterForumEvent {
+                hosting_account: Signer::address_of(account),
+                proposal_type_info: TypeInfo::type_of<ProposalType>(),
+            },
+        );
+
+        move_to(account, voting_forum);
+    }
+
+    /// Create a proposal with the given parameters
+    ///
+    /// @param voting_forum_address The forum's address where the proposal will be stored.
+    /// @param execution_content The execution content that will be given back at resolution time. This can contain
+    /// data such as a capability resource used to scope the execution.
+    /// @param min_vote_threshold The minimum number of votes needed to consider this proposal successful.
+    /// @param expiration_secs The time in seconds at which the proposal expires and can potentially be resolved.
+    /// @return The proposal id.
+    public fun create_proposal<ProposalType: store>(
+        voting_forum_address: address,
+        execution_content: ProposalType,
+        execution_hash: Option<vector<u8>>,
+        min_vote_threshold: u64,
+        expiration_secs: u64,
+    ): u64 acquires VotingForum {
+        let voting_forum = borrow_global_mut<VotingForum<ProposalType>>(voting_forum_address);
+        let proposal_id = voting_forum.next_proposal_id;
+        voting_forum.next_proposal_id = voting_forum.next_proposal_id + 1;
+
+        Table::add(&mut voting_forum.proposals, proposal_id, Proposal {
+            execution_content: Option::some<ProposalType>(execution_content),
+            execution_hash,
+            min_vote_threshold,
+            expiration_secs,
+            yes_votes: 0,
+            no_votes: 0,
+            resolved: false,
+        });
+        proposal_id
+    }
+
+    /// Vote on the given proposal.
+    ///
+    /// @param _proof Required so only the governance module that defines ProposalType can initiate voting.
+    ///               This guarantees that voting eligibility and voting power are controlled by the right governance.
+    /// @param voting_forum_address The address of the forum where the proposals are stored.
+    /// @param proposal_id The proposal id.
+    /// @param num_votes Number of votes. Voting power should be calculated by governance.
+    /// @param should_pass Whether the votes are for yes or no.
+    public fun vote<ProposalType: store>(
+        _proof: &ProposalType,
```

Comment:
this looks confusing, we probably should separate it from the real proposal

Reply:
Per offline discussion, this is required so that only the Governance module can submit the votes directly to Voting. This allows a separation of concerns - the governance module would be responsible for voting power calculation and preventing double voting.